flowmachine/flowmachine/core/subscriber_subset.py

```diff
@@ -0,0 +1,186 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+from abc import abstractmethod
+from sqlalchemy.sql import ClauseElement, select, text, column
+from .query import Query
+
+
+class SubscriberSubsetBase(Query):
+    """
+    Base class for the different types of subscriber subsets.
+    """
+
+    @property
+    @abstractmethod
+    def is_proper_subset(self):
+        raise NotImplementedError(
+            f"Class {self.__class__.__name__} does not implement 'is_proper_subset'"
+        )
+
+    @abstractmethod
+    def apply_subset(self, sql, *, subscriber_identifier):
+        raise NotImplementedError(
+            f"Class {self.__class__.__name__} does not implement 'apply_subset'"
+        )
+
+    def _get_query_attrs_for_dependency_graph(self, analyse=False):
+        # This is a stub implementation of this internal method.
+        # It is needed because SubscriberSubsetBase currently
+        # inherits from flowmachine.Query, so we implement just
+        # enough to ensure Query.dependency_graph() doesn't break.
+        attrs = {}
+        attrs["name"] = self.__class__.__name__
+        attrs["stored"] = "N/A"
+        attrs["cost"] = "N/A"
+        attrs["runtime"] = "N/A"
+        return attrs
+
+
+class AllSubscribers(SubscriberSubsetBase):
+    """
+    Represents the subset of all subscribers (i.e., no subsetting at all).
+    """
+
+    is_proper_subset = False
+
+    def _make_query(self):
+        # Return a dummy string representing this subset. This is only needed
+        # because SubscriberSubsetBase currently inherits from Query, but will
+        # eventually be removed.
+        return "<AllSubscribers>"
+
+    def apply_subset(self, sql, *, subscriber_identifier=None):
+        """
+        Return the input query unchanged, since no subsetting is applied.
+
+        Parameters
+        ----------
+        sql : sqlalchemy.sql.ClauseElement
+            The SQL query to which the subset should be applied.
+
+        subscriber_identifier : str
+            This argument is ignored for subsets of type 'AllSubscribers'.
+        """
+        return sql
+
+
+class SubsetFromFlowmachineQuery(SubscriberSubsetBase):
+    """
+    Represents a subset given by a flowmachine query.
+    """
+
+    is_proper_subset = True
+
+    def __init__(self, flowmachine_query):
```

Comment:
Can we expand the docstring to say what this query should contain in terms of columns, etc.

Reply:
👍